File: 0140-word-break-ii/0140-word-break-ii.java
class Solution {
    public void f(int start, String s, Set<String> set, List<String> res, String cur, int n) {
        if (start == n) {
            cur = cur.stripTrailing();
            res.add(cur);
            return;
        }
        
        for (int end = start + 1; end <= n; end++) {
            String str = s.substring(start, end);
            if (set.contains(str)) {
                f(end, s, set, res, cur + str + " ", n);
            }
        }
    }
    
    public List<String> wordBreak(String s, List<String> wordDict) {
        List<String> res = new ArrayList<>();
        Set<String> dictionary = new HashSet<>(wordDict);
        f(0, s, dictionary, res, "", s.length());
        return res;
    }
}

File: 0140-word-break-ii/0140-word-break-ii.cpp
class Solution {
public:
    void f(int i, unordered_set<string>& dict, vector<string>& res, string cur, int& n, string& s) {
        if (i == n) {
            cur.pop_back();
            res.push_back(cur);
            return;
        }
        
        for (int j = i; j < n; j++) {
            string ck = s.substr(i, j - i + 1);
            if (dict.find(ck) != dict.end()) {
                f(j + 1, dict, res, cur + ck + " ", n, s);
            }
        }
    }
    
    vector<string> wordBreak(string s, vector<string>& wordDict) {
        unordered_set<string> dict(wordDict.begin(), wordDict.end());
        
        vector<string> res;
        
        int n = s.size();
        
        f(0, dict, res, "", n, s);
        
        return res;
    }
};